src/common.h

```diff
@@ -138,16 +138,19 @@ enum {
 };
 typedef unsigned int unescape_flags_t;
 
-// Flags for the escape_string() and escape_string() functions. These are only applicable when the
-// escape style is "script" (i.e., STRING_STYLE_SCRIPT).
+// Flags for the escape_string() function. These are only applicable when the escape style is
+// "script" (i.e., STRING_STYLE_SCRIPT).
 enum {
-    /// Escape all characters, including magic characters like the semicolon.
-    ESCAPE_ALL = 1 << 0,
+    /// Do not escape special fish syntax characters like the semicolon. Only escape nonprintable
+    /// characters and backslashes.
+    ESCAPE_NO_PRINTABLES = 1 << 0,
```

Comment:
I would call this "ESCAPE_NONPRINTABLES" or just "ESCAPE_NONPRINT".

"ESCAPE_NO_PRINTABLES" makes it sound like it escapes everything except for printable characters - and a tilde is printable.

Reply:
With that change, I would be okay with merging everything up to [781a5be](https://github.com/fish-shell/fish-shell/pull/9089/commits/781a5be0baecff88d51f84906c757c5e123cbc2c), so we can concentrate on the pager stuff here.

I'm wary of the change to use the unicode control character symbols - how's the font coverage? - so I would like to check that first.

If possible, that should be extracted to its own PR so we can discuss it separately. A quick check shows that the system I'm on at the moment *has* these symbols, but at the font size in my terminal the `␉` is entirely unreadable and "H T" doesn't scream "tab character" to me (for one because vertical tabs basically aren't a thing and so "horizontal tab" isn't something that really springs to mind).

Reply:
thanks for taking a look

> I would call this "ESCAPE_NONPRINTABLES" or just "ESCAPE_NONPRINT".
>		
>	"ESCAPE_NO_PRINTABLES" makes it sound like it escapes everything except for printable characters - and a tilde is printable.

That sounds right to me - in contrast to the default it does not escape printable characters (except for backslash).
So both ESCAPE_NO_PRINTABLES and ESCAPE_NO_TILDE make us escape fewer characters. We can observe that ESCAPE_NO_PRINTABLES implies ESCAPE_NO_TILDE:

	$ set foo \~\;\1
	$ set -S foo
	$foo: set in global scope, unexported, with 1 elements
	$foo[1]: |~;\ca|

So I'd leave it.

> If possible, that should be extracted to its own PR

definitely

Reply:
I agree that ESCAPE_NONPRINT is more intuitive at first but ESCAPE_NO_PRINTABLES is consistent with the fact that 0=ESCAPE_ALL (and with ESCAPE_NO_TILDE).

We could avoid negations by making the default something like `flags = ESCAPE_NONPRINT | ESCAPE_PRINT`
and then call it like `escape_string(str, ESCAPE_NONPRINT)`.
The problem with that is that it's a bit worse for callers who want other flags like ESCAPE_NO_QUOTED, they'd need `escape_string(str, ESCAPE_ALL | ESCAPE_NO_QUOTED)` where `ESCAPE_ALL = ESCAPE_NONPRINT | ESCAPE_PRINT`.

Reply:
So ESCAPE_NO_PRINTABLES makes it not escape printable characters?

Ah, okay, I read that the wrong way around then. Fine, ship it - it being everything up to https://github.com/fish-shell/fish-shell/pull/9089/commits/db7b8151ca6b526280c27a16642c79a6c2bdb697.